src/System.Management.Automation/engine/Modules/ModuleSpecification.cs

```diff
@@ -266,92 +273,58 @@ public static bool TryParse(string input, out ModuleSpecification result)
-                }
-                else
-                {
-                    result = true;
-                }
-                if (result)
-                {
-                    if (x.Guid.HasValue && y.Guid.HasValue)
-                    {
-                        result = x.Guid.Equals(y.Guid);
-                    }
-                }
-                if (result)
-                {
-                    if (x.Version != null && y.Version != null)
-                    {
-                        result = x.Version.Equals(y.Version);
-                    }
-                    else if (x.Version != null || y.Version != null)
-                    {
-                        result = false;
-                    }
-
-                    if (x.MaximumVersion != null && y.MaximumVersion != null)
-                    {
-                        result = x.MaximumVersion.Equals(y.MaximumVersion);
-                    }
-                    else if (x.MaximumVersion != null || y.MaximumVersion != null)
-                    {
-                        result = false;
-                    }
-
-                    if (result && x.RequiredVersion != null && y.RequiredVersion != null)
-                    {
-                        result = x.RequiredVersion.Equals(y.RequiredVersion);
-                    }
-                    else if (result && (x.RequiredVersion != null || y.RequiredVersion != null))
-                    {
-                        result = false;
-                    }
-                }
+                return true;
             }
 
-            return result;
+            return x != null && y != null
+                && String.Equals(x.Name, y.Name, StringComparison.OrdinalIgnoreCase)
+                && Guid.Equals(x.Guid, y.Guid)
+                && Version.Equals(x.RequiredVersion, y.RequiredVersion)
+                && Version.Equals(x.Version, y.Version)
+                && String.Equals(x.MaximumVersion, y.MaximumVersion);
         }
 
+        /// <summary>
+        /// Get a property-based hashcode for a ModuleSpecification object.
+        /// </summary>
+        /// <param name="obj">The module specification for the object.</param>
+        /// <returns>A hashcode that is always the same for any module specification with the same properties.</returns>
+        /// <remarks>
+        /// The algorithm in this method is adapted from a StackOverflow answer from Jon Skeet: https://stackoverflow.com/a/263416/9944203
+        /// </remarks>
         public int GetHashCode(ModuleSpecification obj)
         {
-            int result = 0;
+            if (obj == null)
+            {
+                return 0;
+            }
+
+            // Two ordinary prime numbers for hashing
+            int p = 23;
+            int q = 59;
 
-            if (obj != null)
+            int result = p; 
+
+            foreach (object property in new object[] { obj.Name, obj.Guid, obj.RequiredVersion, obj.Version, obj.MaximumVersion })
```

Comment:
Looking at this I remembered that we have
```c#
internal static int CombineHashCodes(int h1, int h2, int h3, int h4, int h5)
```
in Utils.cs.

Reply:
Well on the one hand it would be nice to deduplicate our code.

On the other hand:
* The original intent of the refactor here was to roll the repeated operations into a loop. `CombineHashCodes()` doesn't have a `params int[] hs` overload, so we'd have to unroll the loop again.
* The implementation in `CombineHashCodes()` still uses XOR. [This StackOverflow answer](https://stackoverflow.com/a/34229665/9944203) discusses the implementation we borrowed from System.Web.Util.HashCodeCombiner and I notice that another answer is Jon Skeet again disrecommending XOR for hashcodes.

Not for this PR, but maybe it's worth:
* Changing `CombineHashCodes()` to use the add/multiply method, and
* Adding a `CombineHashCodes(int h, param int[] hs)` overload

If we did that it would probably then be worth moving the `ModuleSpecificationComparer.GetHashCode()` method over to use `Utils.CombineHashCodes()`.

Reply:
I agree about XOR but we use it very frequently - I propose to don't block the PR and to postpone the change to a special PR where cleanup all code base. Please open new tracking Issue if you agree.

Reply:
For reference. We have also:
```
internal static int SequenceGetHashCode<T>(this IEnumerable<T> xs)

public int GetHashCode(Tuple<ITypeName, TypeResolutionState> obj)

public override int GetHashCode(ICollection<T> obj)
```